otelconf/internal/provider/envprovider.go

```diff
@@ -0,0 +1,117 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package provider contains various providers
+// used to replace variables in configuration files.
+package provider // import "go.opentelemetry.io/contrib/otelconf/internal/provider"
+
+import (
+	"fmt"
+	"os"
+	"regexp"
+	"strings"
+	"time"
+
+	"go.yaml.in/yaml/v3"
+)
+
+const validationPattern = `^[a-zA-Z_][a-zA-Z0-9_]*$`
+
+var validationRegexp = regexp.MustCompile(validationPattern)
+
+func ReplaceEnvVars(input []byte) ([]byte, error) {
+	// start by replacing all $$ that are not followed by a {
+	doubleDollarSigns := regexp.MustCompile(`\$\$([^{$])`)
```

Comment:
Why this is not as a global variable (or why we are not passing the already parsed regexp when doing recursion)? Parsing the regex each time is expensive.

Same comment applies to other `regexp.MustCompile` occurrences.

Reply:
moved to global variable